### PR TITLE
Customizing the Result of JPA Queries with Aggregation Functions Module

### PR DESCRIPTION
**Pull request by zepfred**

### CODE REVIEW

File: persistence-modules/spring-data-jpa-2/src/test/java/com/baeldung/aggregation/SpringDataAggregateIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+		Assert.assertTrue(countYear2018.getYear().equals(2018));
+		Assert.assertEquals(countYear2018.getTotal().longValue(), 2l);
+		CommentCount countYear2017 = commentsYear.get(2);
+		Assert.assertTrue(countYear2017.getYear().equals(2017));
+		Assert.assertEquals(countYear2017.getTotal().longValue(), 1l);
+	}
+
+	@Test
+	public void whenQueryWithAggregation_thenReturnInterfaceResult() {
+		List<ICommentCount> commentsYear = commentRepository.countTotalCommentsByYearInterface();
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-jpa-2/src/test/java/com/baeldung/aggregation/SpringDataAggregateIntegrationTest.java]: `commentsYear` is a bit of an obscure name for this. `commentCountsByYear` might be clearer.

File: persistence-modules/spring-data-jpa-2/README.md [hunk 2]
```diff
@@ -13,3 +13,4 @@
 - [Difference Between save() and saveAndFlush() in Spring Data JPA](https://www.baeldung.com/spring-data-jpa-save-saveandflush)
 - [Derived Query Methods in Spring Data JPA Repositories](https://www.baeldung.com/spring-data-derived-queries)
 - [LIKE Queries in Spring JPA Repositories](https://www.baeldung.com/spring-jpa-like-queries)
+- [Customizing the Result of JPA Queries with Aggregation Functions](https://www.baeldung.com/spring-jpa-aggregation-function-queries)
```
[the remark this answers is not in the document]
Comment (zepfred) [reply persistence-modules/spring-data-jpa-2/README.md]: Ok

File: persistence-modules/spring-data-jpa-2/src/main/java/com/baeldung/aggregation/model/Comment.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+import java.util.Objects;
+
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.ManyToOne;
+
+@Entity
+public class Comment {
+	@Id
```
[the remark this answers is not in the document]
Comment (zepfred) [reply persistence-modules/spring-data-jpa-2/src/main/java/com/baeldung/aggregation/model/Comment.java]: Ok

File: persistence-modules/spring-data-jpa-2/src/main/java/com/baeldung/aggregation/model/Comment.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	public void setPost(Post post) {
+		this.post = post;
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o)
+			return true;
+		if (!(o instanceof Comment))
+			return false;
```
[the remark this answers is not in the document]
Comment (zepfred) [reply persistence-modules/spring-data-jpa-2/src/main/java/com/baeldung/aggregation/model/Comment.java]: Ok

File: persistence-modules/spring-data-jpa-2/src/main/java/com/baeldung/aggregation/model/Post.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	public void setComments(List<Comment> comments) {
+		this.comments = comments;
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o)
+			return true;
+		if (!(o instanceof Post))
+			return false;
```
[the remark this answers is not in the document]
Comment (zepfred) [reply persistence-modules/spring-data-jpa-2/src/main/java/com/baeldung/aggregation/model/Post.java]: Ok

File: persistence-modules/spring-data-jpa-2/src/test/java/com/baeldung/aggregation/SpringDataAggregateIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+		List<CommentCount> commentsYear = commentRepository.countTotalCommentsByYearClass();
+		CommentCount countYear2019 = commentsYear.get(0);
+		Assert.assertTrue(countYear2019.getYear().equals(2019));
+		Assert.assertEquals(countYear2019.getTotal().longValue(), 1l);
+		CommentCount countYear2018 = commentsYear.get(1);
+		Assert.assertTrue(countYear2018.getYear().equals(2018));
+		Assert.assertEquals(countYear2018.getTotal().longValue(), 2l);
+		CommentCount countYear2017 = commentsYear.get(2);
+		Assert.assertTrue(countYear2017.getYear().equals(2017));
+		Assert.assertEquals(countYear2017.getTotal().longValue(), 1l);
```
[the remark this answers is not in the document]
Comment (zepfred) [reply persistence-modules/spring-data-jpa-2/src/test/java/com/baeldung/aggregation/SpringDataAggregateIntegrationTest.java]: Ok

File: persistence-modules/spring-data-jpa-2/src/test/java/com/baeldung/aggregation/SpringDataAggregateIntegrationTest.java (hunk 1 above)
Comment (zepfred) [reply persistence-modules/spring-data-jpa-2/src/test/java/com/baeldung/aggregation/SpringDataAggregateIntegrationTest.java]: Ok

File: persistence-modules/spring-data-jpa-2/src/test/java/com/baeldung/aggregation/SpringDataAggregateIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        Object[] countYear2017 = commentCountsByYear.get(2);
+        assertEquals(countYear2017[0], Integer.valueOf(2017));
+        assertEquals(countYear2017[1], 1l);
+    }
+
+    @Test
+    public void whenQueryWithAggregation_thenReturnCustomResult() {
+        List<CommentCount> commentCountsByYear = commentRepository.countTotalCommentsByYearClass();
+        CommentCount countYear2019 = commentCountsByYear.get(0);
+        assertEquals(countYear2019.getYear(), Integer.valueOf(2019));
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-jpa-2/src/test/java/com/baeldung/aggregation/SpringDataAggregateIntegrationTest.java]: This assertion is backwards. It should be `assertEquals(expected, actual)`

Comment (zepfred) [reply persistence-modules/spring-data-jpa-2/src/test/java/com/baeldung/aggregation/SpringDataAggregateIntegrationTest.java]: I changed the assertions to use assertThat.

**Pull request merged**